18Rhl: Fix Trajekt handling in Rhine Metropoles (fix #7834)

perwestling opened the pull request:
Redesigned the three green Rhine Metropole hexes to be 2+1/1+2
cities in appearance, but work as one city with 3 slots in
practice. This means that a green such hex that has one or
more slots untokened are passable by everyone. And that such
green hex that are fully tokened are passable for the 3
corporations that have a token in any slot on the hex.

To accomplish this there is a special "make neutral/blocked"
procedure at suitable times during the operation round.
I.e. at start/end of OR, and then any of these 3 hexes
are tokened/upgraded.

Also, fixed upgrade of green Rhine Metropolis hexes.

The methods for route handling, bonuses, distance, revenues
has been rewritten to handle these cases.

Also corrected some minor bugs.

perwestling commented:
At this point in time there are 2 games ongoing, both in SR 8.  I think they can be pinned to let them complete. Any finished games can be archived.

This change do break any existing games that has reached green phase and upgraded a Rhine Metropolis.

tobymao commented on lib/engine/game/g_18_rhl/game.rb:
```diff
@@ -623,79 +641,167 @@ def event_remove_tile_block!
+        def illegal_revisit_of_rhine_metropolis_hex?(actual_visits)
+          actual_visits.select { |v| RHINE_METROPOLIS_HEXES.include?(v.hex.name) && v.hex.tile.color == :green }
+                       .group_by { |v| v.hex.name }.reject { |_, e| e.size < 2 }
+                       .any?
```
it should be group_by.any? {...} not reject? any

perwestling replied:
Fixed

tobymao commented on lib/engine/game/g_18_rhl/game.rb:
```diff
@@ -757,8 +863,50 @@ def player_value(player)
+        end
+
+        def get_token(entity, token)
+          return token unless token.nil?
```
if token

perwestling replied:
Fixed

tobymao commented on lib/engine/game/g_18_rhl/game.rb:
```diff
@@ -905,63 +1013,19 @@ def iron_rhine_bonus(stops)
-          end
-          return bonus unless doubles.positive?
+          doubles = revenue_stops.select { |rs| rs[:rhine_metropolis_bonus] }
+          return bonus unless doubles.any?
```
if doubles.empty?

perwestling replied:
Fixed

perwestling commented:
@tobymao: Have now fixed your comments.
Also updated Welcome message to make players aware that 18Rhl needs alpha testing.

Right now there is only two active 18Rhl games, and they are both in the last (9th) OR set. So depending on when this is deployed, pinning those + archive the rest, or archive all, might be the proposed action.
